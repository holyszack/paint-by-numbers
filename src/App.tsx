import { Grid, Paper } from "@material-ui/core";
import React from "react";
import { sourceFilename$ } from "./app/source_filenames";
import { setSourceFile } from "./app/source_files";
import { sourcePallete$ } from "./app/source_palletes";
import { sourcePreviewUrl$ } from "./app/source_preview_urls";
import { DisplayMessages } from "./components/display_messages";
import { ImagePicker } from "./components/image_picker";
import { ImagePreview } from "./components/image_preview";
import { useObservable } from "./services/hooks/use_observable";

export const App = () => {
    const source = {
        "name": useObservable(sourceFilename$),
        "palette": useObservable(sourcePallete$),
        "previewUrl": useObservable(sourcePreviewUrl$),
        "progress": undefined,
    };
    return (
        <Paper>
            <ImagePicker setImage={setSourceFile} />
            <Grid container={true}>
                <Grid item={true} xs={12} sm={6}>
                    <ImagePreview {...source} name="source" />
                </Grid>
                {/* <Grid item={true} xs={12} sm={6}>
                    <ImagePreview {...target} name="target" />
                </Grid> */}
            </Grid>
            <DisplayMessages />
        </Paper >
    );
};
