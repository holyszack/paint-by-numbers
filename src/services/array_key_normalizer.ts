export function arrayKeyNormalizer(a: number[]) {
    return a.map(Math.floor).toString();
} 
