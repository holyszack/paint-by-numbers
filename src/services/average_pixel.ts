import { RGB } from "../types/rgb";

export function averagePixel(pixels: RGB[]): RGB {
    return pixels.reduce(
        (avg: number[], pixel) => avg.map((value, index) => value + pixel[index]),
        Array(4).fill(0) as RGB,
    ).map((value) => Math.floor(value / pixels.length)) as RGB;
};
