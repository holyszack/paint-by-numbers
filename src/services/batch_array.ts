export function batchArray(batchSize: number) {
    return <T>(source: T[]): T[][] => {
        const output =[];
        for(let i = 0; i < source.length; i += batchSize) {
            output.push(source.slice(i, i + batchSize));
        }
        return output;
    };
}
