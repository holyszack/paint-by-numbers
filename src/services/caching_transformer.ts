export function cachingTransformer<T, K, V>({
    cache = new Map<K | T, V | T>(),
    transformer = (a) => a,
    normalizer = (a) => a,
}: {
    "cache"?: {
        "get": (key: K | T) => V | T | undefined,
        "set": (key: K | T, value: V | T) => void;
    },
    "transformer"?: (a: T) => V | T,
    "normalizer"?: (a: T) => K | T,
}) {
    return (item: T) => {
        const key = normalizer(item);
        const found = cache.get(key);
        if (found) {
            return found;
        } else {
            const value = transformer(item);
            cache.set(key, value);
            return value;
        }
    };
}
