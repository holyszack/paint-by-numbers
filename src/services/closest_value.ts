export const closestValue = <T>(set: T[], test: (a: T, b: T) => number) => (item: T): number => {
    let minimumDistance: number = undefined as unknown as number;
    let minimumIndex = 0;
    set.map((value) => test(value, item))
        .forEach((distance, index) => {
            if (distance < minimumDistance || minimumDistance === undefined) {
                minimumIndex = index;
                minimumDistance = distance;
            }
        });
    return minimumIndex;
};
