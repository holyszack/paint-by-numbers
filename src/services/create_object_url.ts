export function createObjectUrl(file: any) {
    return URL.createObjectURL(file);
}
