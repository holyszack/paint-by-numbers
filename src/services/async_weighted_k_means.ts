import { Observable } from "rxjs";
import { Palette } from "../types/palette";
import { RGB } from "../types/rgb";
import { WeightedRGB } from "../types/weighted_rgb";
import { batchArray } from "./batch_array";
import { closestValue } from "./closest_value";
import { dotReduce } from "./dot_reduce";
import { euclideanDistance } from "./euclidean_distance";
import { generateRandomPoints } from "./generate_random_points";
import { histogramToWeightedRgbs } from "./histogram_to_weighted_rgbs";
import { memoize } from "./memoize";
import { pointSort } from "./point_sort";
import { stopwatch } from "./stopwatch";
import { weightedAveragePixel } from "./weighted_average_pixel";

export type WeightedKMeans = {
    "progress": number;
    "partitions": number;
    "complete": boolean;
    "palette": RGB[];
}
const batchSize = 2500;

export function asyncWeightedKMeans(partitions: number) {
    return (histogram: Map<string, number>) => new Observable<WeightedKMeans>((subscriber) => {
        try {
            const population = histogramToWeightedRgbs(histogram);
            console.log(population.length);
            let palette = generateRandomPoints({ "length": partitions })
                .sort(pointSort());
            subscriber.next({ progress: 0, partitions, complete: false, palette });
            let targets: Palette = [];
            const populations = batchArray(batchSize)(population);
            const kMeanCycle = () => {
                const times = stopwatch();
                targets = palette;
                const paletteMap = memoize(closestValue(targets, euclideanDistance));
                const subPalettes = populations
                    .map((pop) => {
                        return new Promise<WeightedRGB[][]>((resolve, reject) => {
                            setTimeout(() => resolve(pop.reduce(
                                (acc, point) => {
                                    acc[paletteMap(point.value)].push(point);
                                    return acc;
                                },
                                Array.from({ "length": partitions }).map(() => []) as WeightedRGB[][]
                            )), 0);
                        });
                    });
                times.lap("subPalettes");
                Promise.all(subPalettes).then((palettes) => {
                    times.lap("subPalettes done");
                    const tempPalette = dotReduce<WeightedRGB[]>((a, b) => a.concat(b))(palettes);
                    times.lap("tempPalette");
                    const otherPalette = tempPalette
                        .map(weightedAveragePixel);
                    times.lap("otherPalette");
                    palette = otherPalette
                        .sort(pointSort());
                    times.lap("palette");
                    const progress = targets
                        .filter((target, index) => target.toString() === palette[index].toString())
                        .length;
                    times.lap("progress");
                    console.log(times.durations);
                    subscriber.next({ progress, partitions, complete: false, palette });
                    if (progress === partitions) {
                        subscriber.next({ progress: partitions, partitions, complete: true, palette });
                        subscriber.complete();
                    } else {
                        setTimeout(kMeanCycle, 0);
                    }
                });

            };
            kMeanCycle();
        } catch (e) {
            subscriber.error(e);
            subscriber.complete();
        }
    });
}
