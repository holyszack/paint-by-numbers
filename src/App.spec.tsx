import { render } from "@testing-library/react";
import React from "react";
import { App } from "./App";
import { logger } from "./services/logger";

jest.mock("./services/logger");

describe("App", () => {
    beforeEach(() => {
        (logger as jest.Mock).mockReset();
    });
    it("renders without crashing", () => {
        expect(render(<App />).container).toMatchSnapshot();
    });
});
