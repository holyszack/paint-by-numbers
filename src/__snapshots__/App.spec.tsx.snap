// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <input
        accept="image/png"
        type="file"
      />
    </div>
    <span
      class="MuiSlider-root MuiSlider-colorPrimary"
    >
      <span
        class="MuiSlider-rail"
      />
      <span
        class="MuiSlider-track"
        style="left: 0%; width: 48.38709677419355%;"
      />
      <input
        type="hidden"
        value="32"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="0"
        style="left: 0%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="1"
        style="left: 1.6129032258064515%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="2"
        style="left: 3.225806451612903%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="3"
        style="left: 4.838709677419355%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="4"
        style="left: 6.451612903225806%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="5"
        style="left: 8.064516129032258%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="6"
        style="left: 9.67741935483871%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="7"
        style="left: 11.290322580645162%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="8"
        style="left: 12.903225806451612%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="9"
        style="left: 14.516129032258064%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="10"
        style="left: 16.129032258064516%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="11"
        style="left: 17.741935483870968%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="12"
        style="left: 19.35483870967742%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="13"
        style="left: 20.967741935483872%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="14"
        style="left: 22.580645161290324%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="15"
        style="left: 24.193548387096776%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="16"
        style="left: 25.806451612903224%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="17"
        style="left: 27.419354838709676%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="18"
        style="left: 29.032258064516128%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="19"
        style="left: 30.64516129032258%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="20"
        style="left: 32.25806451612903%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="21"
        style="left: 33.87096774193548%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="22"
        style="left: 35.483870967741936%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="23"
        style="left: 37.096774193548384%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="24"
        style="left: 38.70967741935484%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="25"
        style="left: 40.32258064516129%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="26"
        style="left: 41.935483870967744%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="27"
        style="left: 43.54838709677419%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="28"
        style="left: 45.16129032258065%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="29"
        style="left: 46.774193548387096%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive"
        data-index="30"
        style="left: 48.38709677419355%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="31"
        style="left: 50%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="32"
        style="left: 51.61290322580645%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="33"
        style="left: 53.225806451612904%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="34"
        style="left: 54.83870967741935%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="35"
        style="left: 56.45161290322581%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="36"
        style="left: 58.064516129032256%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="37"
        style="left: 59.67741935483871%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="38"
        style="left: 61.29032258064516%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="39"
        style="left: 62.903225806451616%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="40"
        style="left: 64.51612903225806%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="41"
        style="left: 66.12903225806451%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="42"
        style="left: 67.74193548387096%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="43"
        style="left: 69.35483870967742%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="44"
        style="left: 70.96774193548387%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="45"
        style="left: 72.58064516129032%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="46"
        style="left: 74.19354838709677%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="47"
        style="left: 75.80645161290323%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="48"
        style="left: 77.41935483870968%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="49"
        style="left: 79.03225806451613%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="50"
        style="left: 80.64516129032258%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="51"
        style="left: 82.25806451612904%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="52"
        style="left: 83.87096774193549%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="53"
        style="left: 85.48387096774194%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="54"
        style="left: 87.09677419354838%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="55"
        style="left: 88.70967741935483%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="56"
        style="left: 90.3225806451613%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="57"
        style="left: 91.93548387096774%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="58"
        style="left: 93.54838709677419%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="59"
        style="left: 95.16129032258064%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="60"
        style="left: 96.7741935483871%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="61"
        style="left: 98.38709677419355%;"
      />
      <span
        class="MuiSlider-mark"
        data-index="62"
        style="left: 100%;"
      />
      <span
        aria-labelledby="discrete-slider-always"
        aria-orientation="horizontal"
        aria-valuemax="64"
        aria-valuemin="2"
        aria-valuenow="32"
        aria-valuetext="32"
        class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-open-2 PrivateValueLabel-thumb-1"
        data-index="0"
        role="slider"
        style="left: 48.38709677419355%;"
        tabindex="0"
      >
        <span
          class="PrivateValueLabel-offset-3 MuiSlider-valueLabel"
        >
          <span
            class="PrivateValueLabel-circle-4"
          >
            <span
              class="PrivateValueLabel-label-5"
            >
              32
            </span>
          </span>
        </span>
      </span>
    </span>
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <img
            alt="source"
            src="placeholder.png"
            style="max-width: 100%;"
          />
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
